hyperspy/roi.py

```diff
@@ -762,26 +766,31 @@ def __call__(self, signal, out=None, axes=None):
                 shape.append(1)
         mask = mask.reshape(shape)
         mask = np.tile(mask, tiles)
-
         if axes[0].navigate:
-            if len(axes) == 2 and not axes[1].navigate:
+            slicer = signal.inav.__getitem__
+            slices = slices[0:signal.axes_manager.navigation_dimension]
+        else:
+            if len(axes) == 2 and axes[1].navigate:
                 # Special case, since we can no longer slice axes in different
                 # spaces together.
                 if out is None:
-                    return signal.inav[slices[0]].isig[slices[1]]
+                    roi = signal.inav[slices[0]].isig[slices[1]]
                 else:
                     intermediate = signal.inav[slices[0]]
                     intermediate.isig.__getitem__(slices[1], out=out)
+                    roi = out
+                roi.data = np.ma.masked_array(roi.data, mask, hard_mask=True)
+                if out is None:
+                    return roi
+                else:
+                    out.events.data_changed.trigger(out)
                     return
```

Comment:
Looking at this code again, it would probably be better just to say:
```python
slicer = signal.inav[slices.pop(0)].isig.__getitem__
```
Elegant and concise, and avoids the bad practice of having multiple return statements with repeated code (bad practice since you always have to mirror changes to the code across all identical patches, as was forgotten here).